src/frontends/onnx/frontend/src/op/gather_nd.cpp

```diff
@@ -8,19 +8,79 @@
 #include "openvino/op/gather_nd.hpp"
 
 #include "core/operator_set.hpp"
+#include "openvino/op/broadcast.hpp"
+#include "openvino/op/concat.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/op/gather.hpp"
+#include "openvino/op/maximum.hpp"
+#include "openvino/op/shape_of.hpp"
+#include "openvino/op/slice.hpp"
+
 using namespace ov::op;
 
 namespace ov {
 namespace frontend {
 namespace onnx {
 namespace ai_onnx {
 namespace opset_1 {
+
+namespace {
+// Helper function to extract a dimension from a shape tensor at given index
+ov::Output<ov::Node> get_dimension(const ov::Output<ov::Node>& shape, int64_t index) {
+    auto axis = v0::Constant::create(ov::element::i64, ov::Shape{1}, {0});
+    auto start = v0::Constant::create(ov::element::i64, ov::Shape{1}, {index});
+    auto stop = v0::Constant::create(ov::element::i64, ov::Shape{1}, {index + 1});
+    auto step = v0::Constant::create(ov::element::i64, ov::Shape{1}, {1});
+    return std::make_shared<v8::Slice>(shape, start, stop, step, axis);
+}
+}  // namespace
+
 ov::OutputVector gather_nd(const ov::frontend::onnx::Node& node) {
     const ov::OutputVector ng_inputs{node.get_ov_inputs()};
-    const auto data = ng_inputs.at(0);
-    const auto indices = ng_inputs.at(1);
+    auto data = ng_inputs.at(0);
+    auto indices = ng_inputs.at(1);
     const auto batch_dims = node.get_attribute_value<int64_t>("batch_dims", 0);
 
+    // If batch_dims > 0, we need to handle broadcasting for batch dimensions
+    // This is a workaround for ONNXRuntime's non-standard behavior that allows
+    // dimension 1 to broadcast to any size N in batch dimensions
+    if (batch_dims > 0) {
```

Comment:
Do we really need to do broadcast always when batch_dims are provided? I suspect in some cases when shapes are already the same we don't need that? Can we check that case and not insert additional nodes in that case?

Reply:
ok. I fixed it